### PR TITLE
fix(observability): make the shared observability lock reentrant

### PR DESCRIPTION
**Pull request by andyluo7**
## Summary

`lmcache/utils.py` defines one module-global, non-reentrant lock shared by all 30
`@thread_safe` methods in `observability.py`:

```python
_shared_observability_lock = threading.Lock()

def thread_safe(func):
    def wrapper(*args, **kwargs):
        with _shared_observability_lock:      # acquire
            result = func(*args, **kwargs)    # inside
```

`MemoryObj.__del__` (`memory_management.py:672`) calls `parent_allocator.free(self)`, and the
paged allocator's `free()` calls two `@thread_safe` methods —
`update_local_cache_usage` and `update_active_memory_objs_count`
(`paged_tensor_memory_allocator.py:257-258`).

Because the garbage collector can run `__del__` at any allocation point, that update can land
on a thread **already holding** the lock, which hangs permanently. It is reachable in normal
operation through `PrometheusLogger.GetOrCreate` (itself `@thread_safe`), which allocates via
`prometheus_client` while binding label views. Stack captured with py-spy on a hung run:

```
GetOrCreate (observability.py:1863)          <- holds the lock (utils.py:700)
  _create_label_view (observability.py:1920)
  _bind_dynamic_metric_children (observability.py:1493)
  labels (prometheus_client/metrics.py:176)
  <genexpr> (prometheus_client/metrics.py:176)
  __del__ (memory_management.py:688)         <- collector runs here
  free (mixed_memory_allocator.py:201)
  free (paged_tensor_memory_allocator.py:257)
  wrapper (utils.py:699)                     <- re-acquires. Deadlock.
```

Both the holding frame and the acquiring frame are in the **same thread's** stack, so this is
same-thread re-entry, not lock ordering across threads.

**Re-entry is safe here.** The two methods reached through `free()` each assign a single
`LMCStatsMonitor` attribute. They touch no `PrometheusLogger` state and do not iterate the
collector dict that `_bind_dynamic_metric_children` is walking, so allowing re-entry breaks no
invariant.

## Reproducers

Minimal — needs no GPU, hangs with `CUDA_VISIBLE_DEVICES=""`:

```python
from lmcache.utils import thread_safe

@thread_safe
def inner(): return 1

@thread_safe
def outer(): return inner()

outer()   # hangs forever on dev
```

The real path, public APIs only:

```python
import torch
from lmcache.utils import thread_safe
from lmcache.v1.memory_allocators.mixed_memory_allocator import MixedMemoryAllocator
from lmcache.v1.memory_management import MemoryFormat

shape = torch.Size([2, 16, 8, 64])
alloc = MixedMemoryAllocator(size=256 * 1024 * 1024, use_paging=True,
                             shapes=[shape], dtypes=[torch.float16],
                             fmt=MemoryFormat.KV_2TD)
holder = {"obj": alloc.allocate(shape, torch.float16, fmt=MemoryFormat.KV_2TD)}

@thread_safe
def locked_region():
    holder["obj"] = None   # __del__ -> free -> update_local_cache_usage

locked_region()   # hangs forever on dev
```

## How it surfaced

The unit suite hangs at `tests/v1/test_remote_mla_worker_id_as0.py:91` (a `LocalCPUBackend`
constructor) with an identical stack each time. It hit **3 of the 5 runs in which that test
executed** — found while running the suite 8 times to validate an unrelated CI change.

Seen on MI300X/gfx942 with torch 2.10.0+rocm7.0 and MI350X/gfx950 with torch 2.11.0+rocm7.2 —
two architectures, two toolchains, two hosts. I have not reproduced it on NVIDIA, but nothing
in this path is vendor-specific: it is pure Python, and the minimal reproducer needs no GPU.

It is a permanent hang, not a slow test, and there is no per-test timeout
(`pytest-timeout` is not in `requirements/test.txt`), so a job burns its full
`timeout_in_minutes: 60` and reports a timeout naming no test — which reads as an agent
problem and invites a retry rather than an investigation.

## Test plan

The regression test runs the nested call in a worker thread, so a future regression **fails
the test instead of hanging the suite**:

- with this fix: passes in 2.6s
- with the fix reverted to `threading.Lock()`: fails in 32.6s (the 30s wait)

Also on an MI300X:

- 434 passed across `tests/test_observability.py`, `tests/v1/mp_observability`,
  `tests/v1/mp_coordinator/test_observability.py`, `tests/v1/distributed/test_l1_l2_state_metrics.py`
- full unit suite: 9 failed / 5881 passed / 239 skipped vs 10 failed / 5879 passed / 239 skipped
  without the fix, on the same workspace and toolchain. The failure sets differ by one entry, the
  already-flaky `TestDiscoverApiRouters` member that oscillates run to run. Runtime 15:06 vs 14:58.
- `test_remote_mla_worker_id_as0` **passes** in that run

## Note on scope

This is the stop-gap, not the design fix. Keeping `__del__` out of observability entirely
(deferring the stats update rather than taking a lock under the collector), or splitting the
lock so `PrometheusLogger` registration and `LMCStatsMonitor` counters don't share one global,
would remove the hazard rather than tolerate it. Those are judgement calls for the owners of
this area rather than something to decide here.

Adding `pytest-timeout` with a per-test cap is worth doing separately, so a hang surfaces as a
named test rather than an opaque job timeout.

cc @sammshen @ApostaC — you own `memory_management.py`, the other half of this path, though
this diff doesn't touch it.

### CODE REVIEW

**Comment by andyluo7** [comment]
@ApostaC @sammshen @hickeyma @deng451e — when you have a chance, could one of you review this small observability deadlock fix? The PR is mergeable and all current checks pass. Thanks!

**Comment by maobaolong** [comment]
This looks good to me. Switching the shared observability lock to an `RLock` is the right fix here: it preserves the existing serialization while allowing the same-thread re-entry on this path, so I don't see any issue with this change.

**Review by maobaolong — APPROVED** [review]
LGTM

**Comment by andyluo7** [comment]
@ApostaC Could you please take a look? This small deadlock fix is green, including AMD, and already has one approval.

**Comment by andyluo7** [comment]
Rebased onto current `dev`; fresh exact-head CI is green. Ready for required review.

**Comment by andyluo7** [comment]
@hickeyma @sammshen @ApostaC @deng451e — fresh exact-head CI is green after rebase. Could one of you provide the required review?
